### PR TITLE
Split arrow_cast::cast::string into it's own submodule

### LINKED ISSUE #5125: Split up arrow_cast::cast
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The cast kernel is currently implemented as a single large module. This has gotten fairly unwieldy, and should be split up.

**Describe the solution you'd like**

I would like related functionality grouped together into crate private submodules, that are then composed into the top-level cast kernel

Ideas for modules might be:

* `arrow_cast::cast::decimal` - decimal handling logic
* `arrow_cast::cast::list` - list handling logic
* `arrow_cast::cast::dictionary` - dictionary handling logic
* `arrow_cast::cast::string` - string parsing and display logic

In order to make reviews more manageable, and minimize conflicts, these changes should be made across a number of smaller PRs

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by monkwire**
# Which issue does this PR close?

Closes #5125 .

# Rationale for this change
This PR moves the string handling logic from `arrow_cast::cast` into private submodule `arrow_cast::cast::string`, making `arrow_cast::cast` smaller and more manageable.

# What changes are included in this PR?
The string handling functions from `arrow_cast::cast` now live in `arrow_cast::cast::string`, which is located at `arrow-cast/src/cast/string.rs`.

# Are there any user-facing changes?
No

### CODE REVIEW

File: arrow-cast/src/cast/string.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        Some(tz) => {
+            let tz: Tz = tz.as_ref().parse()?;
+            cast_string_to_timestamp_impl(array, &tz, cast_options)?
+        }
+        None => cast_string_to_timestamp_impl(array, &Utc, cast_options)?,
+    };
+    Ok(Arc::new(out.with_timezone_opt(to_tz.clone())))
+}
+
+pub(crate) fn cast_string_to_timestamp_impl<
```
Comment (tustvold) [inline arrow-cast/src/cast/string.rs:91]:
```suggestion
fn cast_string_to_timestamp_impl<
```

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Sorry for the delay, been traveling and was then sick for a few days

**Comment by monkwire** [comment]
No worries at all. Hope you feel better!

File: arrow-cast/src/cast/string.rs (hunk 1 above)
Comment (monkwire) [reply arrow-cast/src/cast/string.rs]: Thanks, I've applied that change in my latest commit.

**Pull request merged**